[plsql] ParseException when parsing EXECUTE IMMEDIATE

**Affects PMD Version:** 6.2 (from net.sourceforge.pmd.eclipse - 4.0.16.v20180412-0833)

**Rule:** parsing a .sql file that has execute immediate with into and using in multiline

**Description:** 
```
There were processing errors!
test.sql: Error while parsing test.sql net.sourceforge.pmd.PMDException: Error while parsing test.sql
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:98)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:50)
	at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:75)
	at net.sourceforge.pmd.processor.MonoThreadProcessor.runAnalysis(MonoThreadProcessor.java:29)
	at net.sourceforge.pmd.processor.AbstractPMDProcessor.processFiles(AbstractPMDProcessor.java:111)
	at net.sourceforge.pmd.eclipse.runtime.cmd.BaseVisitor.reviewResource(BaseVisitor.java:366)
	at net.sourceforge.pmd.eclipse.runtime.cmd.ResourceVisitor.visit(ResourceVisitor.java:30)
	at org.eclipse.core.internal.resources.Resource$2.visit(Resource.java:120)
	at org.eclipse.core.internal.resources.Resource$1.visitElement(Resource.java:84)
	at org.eclipse.core.internal.watson.ElementTreeIterator.doIteration(ElementTreeIterator.java:82)
	at org.eclipse.core.internal.watson.ElementTreeIterator.iterate(ElementTreeIterator.java:129)
	at org.eclipse.core.internal.resources.Resource.accept(Resource.java:94)
	at org.eclipse.core.internal.resources.Resource.accept(Resource.java:52)
	at org.eclipse.core.internal.resources.Resource.accept(Resource.java:117)
	at org.eclipse.core.internal.resources.Resource.accept(Resource.java:105)
	at net.sourceforge.pmd.eclipse.runtime.cmd.ReviewCodeCmd.processResource(ReviewCodeCmd.java:481)
	at net.sourceforge.pmd.eclipse.runtime.cmd.ReviewCodeCmd.processResources(ReviewCodeCmd.java:431)
	at net.sourceforge.pmd.eclipse.runtime.cmd.ReviewCodeCmd.execute(ReviewCodeCmd.java:221)
	at net.sourceforge.pmd.eclipse.runtime.cmd.JobCommandProcessor$1.run(JobCommandProcessor.java:93)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:55)
Caused by: net.sourceforge.pmd.lang.plsql.ast.ParseException: Encountered " <CHARACTER_LITERAL> "\'P\' "" at line 8, column 15.
Was expecting one of:
    "/" ...
    ";" ...
    "=" ...
    "!" ...
    "<" ...
    ">" ...
    "+" ...
    "-" ...
    "||" ...
    "**" ...
    "*" ...
    "AND" ...
    "BETWEEN" ...
    "FROM" ...
    "IN" ...
    "IS" ...
    "LIKE" ...
    "MEMBER" ...
    "MOD" ...
    "NOT" ...
    "OR" ...
    "MULTISET" ...
    "SUBMULTISET" ...
    
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.generateParseException(PLSQLParser.java:40789)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.jj_consume_token(PLSQLParser.java:40643)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.UnlabelledStatement(PLSQLParser.java:9691)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.Statement(PLSQLParser.java:9990)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.ProgramUnit(PLSQLParser.java:5948)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.Global(PLSQLParser.java:496)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.Input(PLSQLParser.java:194)
	at net.sourceforge.pmd.lang.plsql.PLSQLParser.parse(PLSQLParser.java:47)
	at net.sourceforge.pmd.SourceCodeProcessor.parse(SourceCodeProcessor.java:110)
	at net.sourceforge.pmd.SourceCodeProcessor.processSource(SourceCodeProcessor.java:180)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:95)
	... 19 more
```


**Code Sample demonstrating the issue:**
```
CREATE OR REPLACE PROCEDURE test ( p_num_reg OUT number )
 AS
    v_query clob;
 BEGIN
        v_query:='select count(1) from test_tbl where id =:param';

        execute immediate v_query into p_num_reg
        USING 'P';
  END test;
```

Removing the linebreak in the execute immediate bypass the bug:
```
CREATE OR REPLACE PROCEDURE test ( p_num_reg OUT number )
 AS
    v_query clob;
 BEGIN
        v_query:='select count(1) from test_tbl where id =:param';

        execute immediate v_query into p_num_reg USING 'P';
END test;
```


**Running PMD through:** *[Other - Eclipse Plugin]*

Another example with FOR over a select
```
CREATE OR REPLACE PROCEDURE test2 ( a OUT number, b OUT number )
 AS
 BEGIN
        FOR registro IN ( SELECT  SUM(field1),
                                  MAX(field2)
                            INTO a,b
                            FROM  test_tbl
                            GROUP BY fieldx)
        LOOP
            null;
        END LOOP;
END test2;
```

```
There were processing errors!
test2.sql: Error while parsing test2.sql net.sourceforge.pmd.PMDException: Error while parsing test2.sql
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:98)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:50)
	at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:75)
	at net.sourceforge.pmd.processor.MonoThreadProcessor.runAnalysis(MonoThreadProcessor.java:29)
	at net.sourceforge.pmd.processor.AbstractPMDProcessor.processFiles(AbstractPMDProcessor.java:111)
	at net.sourceforge.pmd.eclipse.runtime.cmd.BaseVisitor.reviewResource(BaseVisitor.java:366)
	at net.sourceforge.pmd.eclipse.runtime.cmd.ResourceVisitor.visit(ResourceVisitor.java:30)
	at org.eclipse.core.internal.resources.Resource$2.visit(Resource.java:120)
	at org.eclipse.core.internal.resources.Resource$1.visitElement(Resource.java:84)
	at org.eclipse.core.internal.watson.ElementTreeIterator.doIteration(ElementTreeIterator.java:82)
	at org.eclipse.core.internal.watson.ElementTreeIterator.iterate(ElementTreeIterator.java:129)
	at org.eclipse.core.internal.resources.Resource.accept(Resource.java:94)
	at org.eclipse.core.internal.resources.Resource.accept(Resource.java:52)
	at org.eclipse.core.internal.resources.Resource.accept(Resource.java:117)
	at org.eclipse.core.internal.resources.Resource.accept(Resource.java:105)
	at net.sourceforge.pmd.eclipse.runtime.cmd.ReviewCodeCmd.processResource(ReviewCodeCmd.java:481)
	at net.sourceforge.pmd.eclipse.runtime.cmd.ReviewCodeCmd.processResources(ReviewCodeCmd.java:431)
	at net.sourceforge.pmd.eclipse.runtime.cmd.ReviewCodeCmd.execute(ReviewCodeCmd.java:221)
	at net.sourceforge.pmd.eclipse.runtime.cmd.JobCommandProcessor$1.run(JobCommandProcessor.java:93)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:55)
Caused by: net.sourceforge.pmd.lang.plsql.ast.ParseException: Encountered " "MAX" "MAX "" at line 6, column 35.
Was expecting:
    ")" ...
    
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.generateParseException(PLSQLParser.java:40789)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.jj_consume_token(PLSQLParser.java:40643)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.PrimaryPrefix(PLSQLParser.java:20307)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.PrimaryExpression(PLSQLParser.java:19986)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.IsOfTypeCondition(PLSQLParser.java:19460)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.IsNullCondition(PLSQLParser.java:19126)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.UnaryExpressionNotPlusMinus(PLSQLParser.java:18834)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.UnaryExpression(PLSQLParser.java:18553)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.MultiplicativeExpression(PLSQLParser.java:18226)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.AdditiveExpression(PLSQLParser.java:18112)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.RelationalExpression(PLSQLParser.java:17883)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.EqualityExpression(PLSQLParser.java:17813)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.ConditionalAndExpression(PLSQLParser.java:17765)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.ConditionalOrExpression(PLSQLParser.java:17717)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.Expression(PLSQLParser.java:16778)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.ForStatement(PLSQLParser.java:11611)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.UnlabelledStatement(PLSQLParser.java:9394)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.Statement(PLSQLParser.java:9990)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.ProgramUnit(PLSQLParser.java:5948)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.Global(PLSQLParser.java:496)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.Input(PLSQLParser.java:194)
	at net.sourceforge.pmd.lang.plsql.PLSQLParser.parse(PLSQLParser.java:47)
	at net.sourceforge.pmd.SourceCodeProcessor.parse(SourceCodeProcessor.java:110)
	at net.sourceforge.pmd.SourceCodeProcessor.processSource(SourceCodeProcessor.java:180)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:95)
	... 19 more

```

--- a/docs/pages/release_notes.md
+++ b/docs/pages/release_notes.md
@@ -28,6 +28,8 @@ This is a minor release.
     *   [#1216](https://github.com/pmd/pmd/issues/1216): \[java] UnnecessaryFullyQualifiedName false positive for the same name method
 *   java-design
     *   [#1217](https://github.com/pmd/pmd/issues/1217): \[java] CyclomaticComplexityRule counts ?-operator twice
+*   plsql
+    *   [#1047](https://github.com/pmd/pmd/issues/1047): \[plsql] ParseException when parsing EXECUTE IMMEDIATE
 
 ### API Changes
 

--- a/pmd-plsql/etc/grammar/PldocAST.jjt
+++ b/pmd-plsql/etc/grammar/PldocAST.jjt
@@ -939,7 +939,7 @@ void Skip2NextTerminator(String initiator,String terminator) :
     t = getToken(1);
 	  if(t.image.equals(initiator)) count++;
 	  if(t.image.equals(terminator)) count--;
-    if((null != t.specialToken && beginToken.kind != SELECT && beginToken.kind != INSERT && beginToken.kind != UPDATE && beginToken.kind != DELETE && beginToken.kind != MERGE) || t.kind == EOF)
+    if((null != t.specialToken && beginToken.kind != SELECT && beginToken.kind != INSERT && beginToken.kind != UPDATE && beginToken.kind != DELETE && beginToken.kind != MERGE && beginToken.kind != EXECUTE) || t.kind == EOF)
         return;
     if (t.specialToken != null && "/".equals(t.image))
 	    return;

--- a/pmd-plsql/src/test/java/net/sourceforge/pmd/lang/plsql/ast/ExecuteImmediateTest.java
+++ b/pmd-plsql/src/test/java/net/sourceforge/pmd/lang/plsql/ast/ExecuteImmediateTest.java
@@ -1,0 +1,28 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.plsql.ast;
+
+import org.apache.commons.io.IOUtils;
+import org.junit.Assert;
+import org.junit.Test;
+
+import net.sourceforge.pmd.lang.plsql.AbstractPLSQLParserTst;
+
+public class ExecuteImmediateTest extends AbstractPLSQLParserTst {
+
+    @Test
+    public void parseExecuteImmediate1047a() throws Exception {
+        String code = IOUtils.toString(this.getClass().getResourceAsStream("ExecuteImmediate1047a.pls"));
+        ASTInput input = parsePLSQL(code);
+        Assert.assertNotNull(input);
+    }
+
+    @Test
+    public void parseExecuteImmediate1047b() throws Exception {
+        String code = IOUtils.toString(this.getClass().getResourceAsStream("ExecuteImmediate1047b.pls"));
+        ASTInput input = parsePLSQL(code);
+        Assert.assertNotNull(input);
+    }
+}

--- a/pmd-plsql/src/test/resources/net/sourceforge/pmd/lang/plsql/ast/ExecuteImmediate1047a.pls
+++ b/pmd-plsql/src/test/resources/net/sourceforge/pmd/lang/plsql/ast/ExecuteImmediate1047a.pls
@@ -1,0 +1,14 @@
+--
+-- BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+--
+
+CREATE OR REPLACE PROCEDURE test ( p_num_reg OUT number )
+AS
+    v_query clob;
+BEGIN
+        v_query:='select count(1) from test_tbl where id =:param';
+
+        -- Note: execute immediate statement on multiple lines!
+        execute immediate v_query into p_num_reg
+        USING 'P';
+END test;

--- a/pmd-plsql/src/test/resources/net/sourceforge/pmd/lang/plsql/ast/ExecuteImmediate1047b.pls
+++ b/pmd-plsql/src/test/resources/net/sourceforge/pmd/lang/plsql/ast/ExecuteImmediate1047b.pls
@@ -1,0 +1,12 @@
+--
+-- BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+--
+
+CREATE OR REPLACE PROCEDURE test ( p_num_reg OUT number )
+AS
+    v_query clob;
+BEGIN
+        v_query:='select count(1) from test_tbl where id =:param';
+
+        execute immediate v_query into p_num_reg USING 'P';
+END test;

